File: analysis/analyze.py
#!/usr/bin/env python3
"""
Statistical analysis for zigttp benchmark suite.
Produces: percentiles, confidence intervals, Mann-Whitney U test, markdown report.
"""

import argparse
import json
from pathlib import Path
from datetime import datetime
from typing import Any

try:
    import numpy as np
    from scipy import stats
    HAS_SCIPY = True
except ImportError:
    HAS_SCIPY = False
    print("Warning: scipy not installed, statistical tests disabled")


def load_results(results_dir: Path) -> dict[str, Any]:
    """Load all JSON result files from directory."""
    results = {}
    for f in results_dir.glob("*.json"):
        if f.name == "versions.json":
            results["versions"] = json.loads(f.read_text())
        elif f.name == "system_info.json":
            results["system"] = json.loads(f.read_text())
        elif f.name.startswith("http_"):
            if "http" not in results:
                results["http"] = {}
            data = json.loads(f.read_text())
            key = f"{data.get('runtime', 'unknown')}_{data.get('endpoint', '').replace('/', '_')}"
            results["http"][key] = data
        elif f.name.startswith("microbench_"):
            if "microbench" not in results:
                results["microbench"] = {}
            data = json.loads(f.read_text())
            results["microbench"][data.get("runtime", "unknown")] = data
        elif f.name.startswith("coldstart_"):
            if "coldstart" not in results:
                results["coldstart"] = {}
            data = json.loads(f.read_text())
            results["coldstart"][data.get("runtime", "unknown")] = data
        elif f.name.startswith("memory_"):
            if "memory" not in results:
                results["memory"] = {}
            data = json.loads(f.read_text())
            results["memory"][data.get("runtime", "unknown")] = data
    return results


def compute_percentiles(data: list[float]) -> dict[str, float]:
    """Compute standard percentiles."""
    if not data:
        return {}
    arr = np.array(data) if HAS_SCIPY else data
    if HAS_SCIPY:
        return {
            "mean": float(np.mean(arr)),
            "median": float(np.median(arr)),
            "std": float(np.std(arr)),
            "p50": float(np.percentile(arr, 50)),
            "p95": float(np.percentile(arr, 95)),
            "p99": float(np.percentile(arr, 99)),
            "min": float(np.min(arr)),
            "max": float(np.max(arr)),
            "n": len(arr)
        }
    else:
        sorted_data = sorted(data)
        n = len(sorted_data)
        return {
            "mean": sum(data) / n,
            "median": sorted_data[n // 2],
            "min": sorted_data[0],
            "max": sorted_data[-1],
            "n": n
        }


def mann_whitney_test(a: list[float], b: list[float]) -> dict[str, Any]:
    """Perform Mann-Whitney U test for comparing distributions."""
    if not HAS_SCIPY or len(a) < 3 or len(b) < 3:
        return {"available": False}
    stat, p_value = stats.mannwhitneyu(a, b, alternative='two-sided')
    return {
        "available": True,
        "statistic": float(stat),
        "p_value": float(p_value),
        "significant_at_005": p_value < 0.05
    }


def generate_report(results: dict[str, Any], output_path: Path) -> None:
    """Generate markdown report from results."""
    lines = []
    lines.append("# zigttp Benchmark Results")
    lines.append("")
    lines.append(f"**Generated:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    lines.append("")

    # System info
    if "system" in results:
        sys = results["system"]
        lines.append("## System Information")
        lines.append("")
        lines.append(f"- **OS:** {sys.get('os', 'unknown')} {sys.get('os_version', '')}")
        lines.append(f"- **CPU:** {sys.get('cpu_model', 'unknown')}")
        lines.append(f"- **Cores:** {sys.get('cpu_cores', 'unknown')}")
        lines.append(f"- **RAM:** {sys.get('ram_gb', 'unknown')} GB")
        lines.append("")

        runtimes = sys.get("runtimes", {})
        if runtimes:
            lines.append("### Runtime Versions")
            lines.append("")
            for rt, ver in runtimes.items():
                lines.append(f"- **{rt}:** {ver}")
            lines.append("")

    # HTTP Results
    if "http" in results:
        lines.append("## HTTP Benchmark Results")
        lines.append("")
        lines.append("| Runtime | Endpoint | RPS | p99 Latency |")
        lines.append("|---------|----------|-----|-------------|")

        for key, data in sorted(results["http"].items()):
            runtime = data.get("runtime", "unknown")
            endpoint = data.get("endpoint", "unknown")
            metrics = data.get("metrics", {})
            rps = metrics.get("requests_per_second", 0)
            p99 = metrics.get("latency_p99_secs", 0)
            lines.append(f"| {runtime} | {endpoint} | {rps:,.0f} | {p99:.4f}s |")
        lines.append("")

    # Microbenchmark Results
    if "microbench" in results:
        lines.append("## Microbenchmark Results")
        lines.append("")

        # Get all benchmark names
        bench_names = set()
        for runtime_data in results["microbench"].values():
            benchmarks = runtime_data.get("benchmarks", {})
            bench_names.update(benchmarks.keys())

        if bench_names:
            runtimes = list(results["microbench"].keys())
            header = "| Benchmark |" + "|".join(f" {rt} |" for rt in runtimes)
            lines.append(header)
            lines.append("|" + "|".join(["---"] * (len(runtimes) + 1)) + "|")

            for bench in sorted(bench_names):
                row = f"| {bench} |"
                for rt in runtimes:
                    benchmarks = results["microbench"][rt].get("benchmarks", {})
                    if bench in benchmarks:
                        ops = benchmarks[bench].get("ops_per_sec", 0)
                        row += f" {ops:,.2f} ops/s |"
                    else:
                        row += " - |"
                lines.append(row)
            lines.append("")

    # Cold Start Results
    if "coldstart" in results:
        lines.append("## Cold Start Results")
        lines.append("")
        lines.append("| Runtime | Mean | Median | p95 | p99 |")
        lines.append("|---------|------|--------|-----|-----|")

        for runtime, data in sorted(results["coldstart"].items()):
            metrics = data.get("metrics", {})
            mean_us = metrics.get("mean_us", 0)
            median_us = metrics.get("median_us", 0)
            p95_us = metrics.get("p95_us", 0)
            p99_us = metrics.get("p99_us", 0)
            lines.append(f"| {runtime} | {mean_us:,}us | {median_us:,}us | {p95_us:,}us | {p99_us:,}us |")
        lines.append("")

    # Memory Results
    if "memory" in results:
        lines.append("## Memory Usage")
        lines.append("")
        lines.append("| Runtime | Baseline | Peak | Avg |")
        lines.append("|---------|----------|------|-----|")

        for runtime, data in sorted(results["memory"].items()):
            metrics = data.get("metrics", {})
            baseline = metrics.get("baseline_kb", 0)
            peak = metrics.get("peak_kb", 0)
            avg = metrics.get("avg_kb", 0)
            lines.append(f"| {runtime} | {baseline:,} KB | {peak:,} KB | {avg:,} KB |")
        lines.append("")

    # Methodology link
    lines.append("## Methodology")
    lines.append("")
    lines.append("See [methodology.md](../methodology.md) for detailed test methodology.")
    lines.append("")

    # Write report
    output_path.write_text("\n".join(lines))
    print(f"Report written to: {output_path}")


def main():
    parser = argparse.ArgumentParser(description="Analyze zigttp benchmark results")
    parser.add_argument("--results-dir", required=True, help="Directory containing result JSON files")
    parser.add_argument("--output", required=True, help="Output markdown file path")
    args = parser.parse_args()

    results_dir = Path(args.results_dir)
    output_path = Path(args.output)

    if not results_dir.exists():
        print(f"Results directory not found: {results_dir}")
        return 1

    results = load_results(results_dir)
    if not results:
        print("No results found")
        return 1

    generate_report(results, output_path)
    return 0


if __name__ == "__main__":
    exit(main())

File: lib/analyze.ts
/**
 * Analysis and report generation
 * Replaces Python analyze.py
 */

import { join } from "https://deno.land/std@0.220.0/path/mod.ts";
import { loadResults, loadBaseline, type SystemInfo } from "./results.ts";
import { formatOpsPerSec, type MicrobenchResult } from "./microbench.ts";
import { type HttpResult } from "./http.ts";
import { type ColdStartResult } from "./coldstart.ts";
import { type MemoryResult } from "./memory.ts";
import { detectMode } from "./utils.ts";

type ParsedResults = {
  system?: SystemInfo;
  http: HttpResult[];
  coldstart: ColdStartResult[];
  memory: MemoryResult[];
  microbench: MicrobenchResult[];
};

/**
 * Parse raw JSON files into structured results
 */
function parseResults(rawResults: Record<string, unknown>): ParsedResults {
  const parsed: ParsedResults = {
    http: [],
    coldstart: [],
    memory: [],
    microbench: [],
  };

  for (const [filename, data] of Object.entries(rawResults)) {
    if (filename === "system_info.json") {
      parsed.system = data as SystemInfo;
    } else if (filename.startsWith("http_")) {
      parsed.http.push(data as HttpResult);
    } else if (filename.startsWith("coldstart_")) {
      parsed.coldstart.push(data as ColdStartResult);
    } else if (filename.startsWith("memory_")) {
      parsed.memory.push(data as MemoryResult);
    } else if (filename.startsWith("microbench_")) {
      parsed.microbench.push(data as MicrobenchResult);
    }
  }

  return parsed;
}

/**
 * Format number with comma separators
 */
function formatNumber(n: number): string {
  return n.toLocaleString();
}

/**
 * Generate markdown report from results
 */
export function generateReport(
  results: ParsedResults,
  baseline?: ParsedResults
): string {
  const lines: string[] = [];

  lines.push("# Benchmark Results");
  lines.push("");
  lines.push(`**Generated:** ${new Date().toISOString().slice(0, 19).replace("T", " ")}`);
  lines.push("");

  // System info
  if (results.system) {
    const sys = results.system;
    lines.push("## System Information");
    lines.push("");
    lines.push(`- **OS:** ${sys.os} ${sys.os_version}`);
    lines.push(`- **CPU:** ${sys.cpu_model}`);
    lines.push(`- **Cores:** ${sys.cpu_cores}`);
    lines.push(`- **RAM:** ${sys.ram_gb} GB`);
    lines.push("");

    if (sys.runtimes && Object.keys(sys.runtimes).length > 0) {
      lines.push("### Runtime Versions");
      lines.push("");
      for (const [rt, ver] of Object.entries(sys.runtimes)) {
        lines.push(`- **${rt}:** ${ver}`);
      }
      lines.push("");
    }
  }

  // HTTP Results
  if (results.http.length > 0) {
    lines.push("## HTTP Benchmark Results");
    lines.push("");

    // Group by endpoint
    const sortedHttp = [...results.http].sort((a, b) => {
      if (a.runtime !== b.runtime) return a.runtime.localeCompare(b.runtime);
      return a.endpoint.localeCompare(b.endpoint);
    });

    lines.push("| Runtime | Endpoint | RPS | p99 Latency |");
    lines.push("|---------|----------|-----|-------------|");

    for (const data of sortedHttp) {
      const rps = data.metrics.requests_per_second;
      const p99 = data.metrics.latency_p99_secs;
      lines.push(`| ${data.runtime} | ${data.endpoint} | ${formatNumber(Math.round(rps))} | ${p99.toFixed(4)}s |`);
    }
    lines.push("");
  }

  // Microbenchmark Results
  if (results.microbench.length > 0) {
    lines.push("## Microbenchmark Results");
    lines.push("");

    // Collect all benchmark names
    const benchNames = new Set<string>();
    for (const mb of results.microbench) {
      for (const name of Object.keys(mb.benchmarks)) {
        benchNames.add(name);
      }
    }

    const runtimes = results.microbench.map((mb) => mb.runtime).sort();
    const sortedNames = [...benchNames].sort();

    if (sortedNames.length > 0 && runtimes.length > 0) {
      const header = `| Benchmark |${runtimes.map((rt) => ` ${rt} |`).join("")}`;
      lines.push(header);
      lines.push("|" + "---|".repeat(runtimes.length + 1));

      for (const name of sortedNames) {
        let row = `| ${name} |`;
        for (const rt of runtimes) {
          const mb = results.microbench.find((m) => m.runtime === rt);
          if (mb && mb.benchmarks[name]) {
            const ops = mb.benchmarks[name].ops_per_sec ?? 0;
            row += ` ${formatOpsPerSec(ops)} ops/s |`;
          } else {
            row += " - |";
          }
        }
        lines.push(row);
      }
      lines.push("");
    }
  }

  // Cold Start Results
  if (results.coldstart.length > 0) {
    lines.push("## Cold Start Results");
    lines.push("");
    lines.push("| Runtime | Mean | Median | p95 | p99 |");
    lines.push("|---------|------|--------|-----|-----|");

    const sortedColdstart = [...results.coldstart].sort((a, b) =>
      a.runtime.localeCompare(b.runtime)
    );

    for (const data of sortedColdstart) {
      const m = data.metrics;
      lines.push(
        `| ${data.runtime} | ${formatNumber(m.mean_us)}us | ${formatNumber(m.median_us)}us | ${formatNumber(m.p95_us)}us | ${formatNumber(m.p99_us)}us |`
      );
    }
    lines.push("");
  }

  // Memory Results
  if (results.memory.length > 0) {
    lines.push("## Memory Usage");
    lines.push("");
    lines.push("| Runtime | Baseline | Peak | Avg |");
    lines.push("|---------|----------|------|-----|");

    const sortedMemory = [...results.memory].sort((a, b) =>
      a.runtime.localeCompare(b.runtime)
    );

    for (const data of sortedMemory) {
      const m = data.metrics;
      // Skip if metrics are missing (broken memory profiling)
      if (m.baseline_kb && m.peak_kb) {
        lines.push(
          `| ${data.runtime} | ${formatNumber(m.baseline_kb)} KB | ${formatNumber(m.peak_kb)} KB | ${formatNumber(m.avg_kb)} KB |`
        );
      } else {
        lines.push(`| ${data.runtime} | - | - | - |`);
      }
    }
    lines.push("");
  }

  // Baseline comparison
  if (baseline) {
    lines.push("## Baseline Comparison (vs Deno)");
    lines.push("");

    // HTTP comparison
    if (results.http.length > 0 && baseline.http.length > 0) {
      lines.push("### HTTP Throughput Ratio");
      lines.push("");

      for (const result of results.http) {
        if (result.runtime === "deno") continue;

        const baselineMatch = baseline.http.find(
          (b) => b.endpoint === result.endpoint
        );
        if (baselineMatch) {
          const ratio = result.metrics.requests_per_second / baselineMatch.metrics.requests_per_second;
          const percent = ((ratio - 1) * 100).toFixed(1);
          const sign = ratio >= 1 ? "+" : "";
          lines.push(`- ${result.endpoint}: ${ratio.toFixed(2)}x (${sign}${percent}%)`);
        }
      }
      lines.push("");
    }

    // Microbench comparison
    if (results.microbench.length > 0 && baseline.microbench.length > 0) {
      const baselineMb = baseline.microbench.find((m) => m.runtime === "deno");
      const resultMb = results.microbench.find((m) => m.runtime !== "deno");

      if (baselineMb && resultMb) {
        lines.push("### Microbenchmark Ratio");
        lines.push("");

        const names = Object.keys(resultMb.benchmarks).sort();
        for (const name of names) {
          const baseOps = baselineMb.benchmarks[name]?.ops_per_sec;
          const resultOps = resultMb.benchmarks[name]?.ops_per_sec;

          if (baseOps && resultOps) {
            const ratio = resultOps / baseOps;
            const percent = ((ratio - 1) * 100).toFixed(1);
            const sign = ratio >= 1 ? "+" : "";
            lines.push(`- ${name}: ${ratio.toFixed(2)}x (${sign}${percent}%)`);
          }
        }
        lines.push("");
      }
    }
  }

  lines.push("## Methodology");
  lines.push("");
  lines.push("See [methodology.md](../methodology.md) for detailed test methodology.");
  lines.push("");

  return lines.join("\n");
}

/**
 * Generate and save report to file
 */
export async function generateReportFile(
  resultsDir: string,
  outputPath?: string
): Promise<void> {
  const rawResults = await loadResults(resultsDir);
  const results = parseResults(rawResults);

  // Try to load baseline for comparison
  const mode = detectMode(resultsDir, rawResults);
  const rawBaseline = await loadBaseline(mode);
  const baseline = rawBaseline ? parseResults(rawBaseline) : undefined;

  const report = generateReport(results, baseline);

  const reportPath = outputPath ?? join(resultsDir, "report.md");
  await Deno.writeTextFile(reportPath, report);

  console.log(`Report written to: ${reportPath}`);
}

/**
 * Compare two result directories
 */
export async function compareResults(
  resultsDir: string,
  baselineDir: string
): Promise<string> {
  const rawResults = await loadResults(resultsDir);
  const rawBaseline = await loadResults(baselineDir);

  const results = parseResults(rawResults);
  const baseline = parseResults(rawBaseline);

  return generateReport(results, baseline);
}
